(baby-project-server) #3 initial version of api

dusan-turajlic opened the pull request:
Created an initial version of the API. Its currenly very simple, There is a websocket broker and a couple of controllers. You can use a library called SockJS to connect to the webscoket connection point. The root of the API looks like this

```javascript
{
  "_links": {
    "self": {
      "href": "http://localhost"
    },
    "ws:connect": {
      "href": "http://localhost/connect"
    }
  },
  "_embedded": {
    "channel": [
      {
        "name": "general",
        "subscription": "/channel/general",
        "_links": {
          "self": {
            "href": "http://localhost/channel/general"
          }
        }
      },
      {
        "name": "random",
        "subscription": "/channel/random",
        "_links": {
          "self": {
            "href": "http://localhost/channel/random"
          }
        }
      }
    ]
  }
}
```

From there you get all the needed information on how to connect to the websocket endoint. The api alos provides channels and all the information you need to sen messages and subscribe to a particular channel, the `subscription` field is there so that the client can subscribe to a specific channel, we should get use to using that field because in the future it might not be the same as the path.

dusan-turajlic committed:
(baby-project-server) initial version of api

AndreasNasman commented on src/main/kotlin/turku/forge/babyproject/config/WebSocketConfig.kt:
```diff
@@ -5,21 +5,26 @@ import org.springframework.messaging.simp.config.MessageBrokerRegistry
-        registry.setApplicationDestinationPrefixes( APPLICATION_DESTINATIONS )
+    override fun registerStompEndpoints(registry: StompEndpointRegistry) {
+        registry.addEndpoint(STOMP_ENDPOINT)
+                // @TODO: get allowed origins from config file
```
I think we should avoid having `TODO`s in our code; it's extra noise and they're often left there indefinitely. They make for perfect issues instead!

dusan-turajlic replied:
Ahh but this is a love letter to my self so I don't forget what I need to do before this implementasion is done

AndreasNasman commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,69 @@
+    }
+
+    private fun getChannels(): List<String> {
+        // @TODO: remove the hard coded channel names once we have a database and a way to create and store channels
```
I think we should avoid having `TODO`s in our code; it's extra noise and they're often left there indefinitely. They make for perfect issues instead!

AndreasNasman commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,69 @@
  * Rest control for the root of the api.
  * if you serve this api under `api.example.com` all the request going to `/`
  * will be routed to this controller.
+ * @TODO: Figure out how to handle the CORS logic
```
I think we should avoid having `TODO`s in our code; it's extra noise and they're often left there indefinitely. They make for perfect issues instead!

AndreasNasman commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,69 @@
-const val API_PATH = "/"
+const val API_PATH = ""
+
+// @TODO: maybe move these relations to their own ENUM if needed
```
I think we should avoid having `TODO`s in our code; it's extra noise and they're often left there indefinitely. They make for perfect issues instead!

AndreasNasman commented on src/main/kotlin/turku/forge/babyproject/api/ChannelController.kt:
```diff
@@ -0,0 +1,40 @@
+}
+
+/**
+ * A POJO for messages
```
What's a POJO?

dusan-turajlic replied:
It stands for `Plain old Java object` I need to still document this stuff a bit better, But this is still a `WIP` I just need some one to take a look at the logic and see if I'm doing anything crazy wrong here

AndreasNasman commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,69 @@
 }
+
+/**
+ * This is the index resource for this file
```
I don't quite follow what this comment is. Is it a TODO or documentation?

dusan-turajlic replied:
Its supose to be initial documentatsion. I need to improve all these docs for all of these. Need to figure out what to write there to. Its hard to document this stuff since I need a good location for all these. 

could essentaly document all the resouces as 
```
/**
 * A <type> resource
 * @see {path.to.base.resource.package}
 */
```

That way there is one location where I can explain what a resouce is.

AndreasNasman commented:
Couldn't comment much on the actual Kotlin code since I don't know that language, but a few minor considerations came to mind. 😄

- Do we want to document every method?
- Should we have newlines at the end of every file?

dusan-turajlic commented:
@AndreasNasman 
> Do we want to document every method?

Because this is an opensource project I think we should. We can document stuff in Wikis all we want but code docs is the place most likely to be kept up to date.

> Should we have newlines at the end of every file?

We should an this does. if it dident there would be a little red sign at the end of the file complaining that there is no new like at the end of file. Github just does not show the extra line.

![image](https://user-images.githubusercontent.com/25513545/53688189-67c9f200-3d40-11e9-829d-0e902fd08fb4.png)

dusan-turajlic committed:
(baby-project-server) Adds support for config file

dusan-turajlic commented:
I updated the PR with the config file changes I just need to add some integration tests and we will be good to go. I can the remove the WIP from this PR

dusan-turajlic committed:
(baby-project-server) Adds integration tests for the controllers

dusan-turajlic commented:
I removed the WIP status But I will still update the documentasion and generaly take a look at what we want to put where.

dusan-turajlic committed:
(baby-project-server) Adds in check for channelController context

dusan-turajlic committed:
(baby-project-server) Updates the documentation

dusan-turajlic commented:
I have now updated the documentation so this PR is ready for a serious review

edrik commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,75 @@
+// Relation constants
+const val WS_RELATION = "ws:connect"
+const val CHANNEL_RELATION = "channel"
+const val INDEX_RELATION = "index"
```
Spring HATEOAS already has this https://docs.spring.io/spring-hateoas/docs/current/api/org/springframework/hateoas/IanaLinkRelations.html#INDEX

dusan-turajlic replied:
Good to know will remove that 👍🏼

edrik commented:
Considering what this API endpoint does so far, this might be of interest:
https://www.asyncapi.com/

edrik commented on src/main/kotlin/turku/forge/babyproject/App.kt:
```diff
@@ -14,5 +16,16 @@ class App
+@Configuration
+@EnableConfigurationProperties
+@ConfigurationProperties(prefix = "spring.config")
+class ConfigProperties(val cors: List<String> = arrayListOf())
```
Is `spring.config` not configurations specific to this app?
If so, name-spacing it a bit less generic might reduce confusion.

dusan-turajlic replied:
I changed it now to just be called `babyproject` that will propobly be changed to something else once the name of the product is actualy choosen 😅

edrik commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,75 @@
+
+// Relation constants
+const val WS_RELATION = "ws:connect"
+const val CHANNEL_RELATION = "channel"
```
This is not a registered IANA Link relation, so basically it should also be prefixed in order to be a valid URI.

Would suggest using a `DefaultCurieProvider`, to add a common prefix to all relations. And the `ws:` prefix in this case is a bit unclear in the sense that it seems to refer to "WebSocket", but that is a really generic concept and not at all app specific.

dusan-turajlic replied:
Just want to be sure I understand correctly. I should define a `DefaultCurieProvider` bean with a default prefix like lets say `bp` short for `baby project`.  And when I create my LinkRelations I should do
```kotlin
val CHANNEL_RELATION = HalLinkRelation.curied(CURIE_NAMESPACE, "channel")
```

when creating the realtion and not use `LinkRelation.of`?

edrik replied:
No, you should still use the normal `LinkRelation.of("channel")`, but the `DefaultCurieProvider` takes care of prefixing every none IANA link relation with the default curie so that the link rel is `bp:channel` and a corresponding `cure` entry is generated for the output.

edrik replied:
But use the source
https://github.com/spring-projects/spring-hateoas/blob/master/src/main/java/org/springframework/hateoas/mediatype/hal/DefaultCurieProvider.java
😁

edrik commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,75 @@
+const val API_PATH = ""
+
+// Relation constants
+const val WS_RELATION = "ws:connect"
```
Using a verb as a link relation is usually a nono, would `endpoint` or `connection-url` be more describing.

dusan-turajlic replied:
Yea you are correct about that. I rember reading about the verb thing at some point. I think somthing like `sockjs-endpoint` would work best in this situation, because its a link to the sockjs endpoint 😅

edrik commented on src/main/kotlin/turku/forge/babyproject/resources/Resource.kt:
```diff
@@ -0,0 +1,33 @@
+ * but we may keep it who knows
+ */
+abstract class Resource<T : Resource<T>> : RepresentationModel<T>() {
+    @JsonProperty("_embedded")
```
Might be worth looking at `HalModelBuilder` to mange embedding in general.

dusan-turajlic replied:
There is actualy a good reason for not using it. Or ok there is a reason. We don't really know if we want to use HAL or something else. We are in the prosses figuring that out. I figured this would be a eayse and simple way of hiding away what is beeing used. Because if we don't use HAL we need to have some other type of Modle bulder for what ever else we end up use. But yea I totlay could have used that for the one request that returns the HAL

edrik reviewed (commented):
Some generic comments on REST concepts and Spring HATEOAS

dusan-turajlic commented:
> Considering what this API endpoint does so far, this might be of interest:
> https://www.asyncapi.com/

@edrik Just to clarify are you suggesting we are use to https://www.asyncapi.com/ to document our flow? Or am I missing the point of asyncapi?

edrik commented:
> > Considering what this API endpoint does so far, this might be of interest:
> > https://www.asyncapi.com/
> 
> @edrik Just to clarify are you suggesting we are use to https://www.asyncapi.com/ to document our flow? Or am I missing the point of asyncapi?

@dusan-turajlic Its a reasonable new endeavor to establish a standardized way to describe a async/pub/sub type API. But it might be a option for describing the websocket parts. Basically linking in the async api description from the api root.
But not really suggesting anything specific with the comment other than that it might be worth looking in to.

dusan-turajlic commented:
> @dusan-turajlic Its a reasonable new endeavor to establish a standardized way to describe a async/pub/sub type API. But it might be a option for describing the websocket parts. Basically linking in the async api description from the api root.
> But not really suggesting anything specific with the comment other than that it might be worth looking in to.

@edrik Thank you for the suggestion was looking in to it yesturday and it seems really neat. Especialy for a opensource where this info needs to be well documented and easaly available.

perlun commented on src/main/kotlin/turku/forge/babyproject/api/RootApiController.kt:
```diff
@@ -1,19 +1,75 @@
-        return "Hello World"
+    /**
+     * Mapping for the root of the API
+     * This is the method that is ran when a request is fired at the root
```
```suggestion
     * This is the method that is run when a request is fired at the root
```

perlun commented on src/main/kotlin/turku/forge/babyproject/config/WebSocketConfig.kt:
```diff
@@ -1,25 +1,37 @@
 
 /**
- * This is a configuration where we extend and override the base `WebSocketMessageBrokerConfigurer`
+ * This class enables the Websockets broker witch
```
Don't go witchhunting in the 21st century. :grin: 

```suggestion
 * This class enables the Websockets broker which
```

dusan-turajlic replied:
True, It sucks that witch is a gramaticly correct word so not even my IDE that checks for spelling and gramatical errors notices when I got witch hunting 🤣

perlun commented on src/main/kotlin/turku/forge/babyproject/resources/Resource.kt:
```diff
@@ -0,0 +1,33 @@
+ * this class serves as a abstraction for the media container we are currently using.
+ *
+ * Its likely that this will be removed once we have chosen a media container
+ * but we may keep it who knows
```
This part of the comment could actually be removed. The statement here holds true for almost all code in any project: it might be removed, it might be kept; stating this is just so obvious that it doesn't add much value, at least not to me. :sweat_smile:

dusan-turajlic replied:
Yea that is true. This entier file has been removed in my upp and comming commit. But I do agree it adds no benefit to state that.

perlun commented on src/test/kotlin/turku/forge/babyproject/api/RootApiControllerTest.kt:
```diff
@@ -0,0 +1,31 @@
+package turku.forge.babyproject.api
+
+import jdk.jfr.Description
```
Hmm, this import seems to be unused? (seems like a rather interesting annotation/class though!)

dusan-turajlic replied:
Yea I have not added any check styles to the code yet so I just need to rember to peress `cmd+alt+L` to get rid of thouse. I have an issue for adding in some code style stuff #7

dusan-turajlic replied:
Yea that is true, I think that might be a remenint from me trying to figure out how to write these tests and what needs to be imported. Intelij just imports everything like magic

dusan-turajlic committed:
(baby-project-server) Adds changes mentiond in code review

I removed the resource diretory and changed to using HalModelBuilder.
No matter what Media Type we end up using there will still be a need for
refactoring so might as well go all in on HAL untill we decide otherwise

dusan-turajlic commented:
I updated the code according to the feedback mention by @edrik and @perlun the response now looks like this 
```javascript
{
  "_embedded": {
    "bp:channel": [
      {
        "name": "general",
        "subscription": "/channel/general",
        "_links": {
          "self": {
            "href": "http://localhost:8080/channel/general"
          }
        }
      },
      {
        "name": "random",
        "subscription": "/channel/random",
        "_links": {
          "self": {
            "href": "http://localhost:8080/channel/random"
          }
        }
      }
    ]
  },
  "_links": {
    "bp:sockjs-endpoint": {
      "href": "http://localhost:8080/connect"
    },
    "self": {
      "href": "http://localhost:8080"
    },
    "curies": [
      {
        "href": "https://docs.turkuforge.fi/{#rel}",
        "name": "bp",
        "templated": true
      }
    ]
  }
}
```
I have no ida why `_embedded` now comes first but it does not make a difference in paracitce.

I added the `DefaultCurieProvider` bean and it worked for adding the `bp:` for any link I add to the resource. But no matter what I tried I could not get it to work when adding embedded content. Migrated to `HalModelBuilder` to see if that was the problem but no I still needed to manualy specify `bp:` to `channel`

edrik commented:
> I added the `DefaultCurieProvider` bean and it worked for adding the `bp:` for any link I add to the resource. But no matter what I tried I could not get it to work when adding embedded content. Migrated to `HalModelBuilder` to see if that was the problem but no I still needed to manualy specify `bp:` to `channel`

That's a bit odd, internaly there is `EmbeddedMapper` that should pick up the `CurieProvider` 🤔

dusan-turajlic commented:
I'll take another look at it tomorrow maybe I just defined the bean wrong or something 🤷🏻‍♂️

dusan-turajlic committed:
(baby-project-server) Removes hardcoded curie from embedded

dusan-turajlic commented:
@edrik I got it working but I think this might be a bug in the hal model. It now looks like this.
```javascript
{
  "_embedded": {
    "bp:channelList": [
      {
        "name": "general",
        "subscription": "/channel/general",
        "_links": {
          "self": {
            "href": "http://localhost:8080/channel/general"
          }
        }
      },
      {
        "name": "random",
        "subscription": "/channel/random",
        "_links": {
          "self": {
            "href": "http://localhost:8080/channel/random"
          }
        }
      }
    ]
  },
  "_links": {
    "bp:sockjs-endpoint": {
      "href": "http://localhost:8080/connect"
    },
    "self": {
      "href": "http://localhost:8080"
    },
    "curies": [
      {
        "href": "https://docs.turkuforge.fi/{#rel}",
        "name": "bp",
        "templated": true
      }
    ]
  }
}
```

Before I specified the LinkRelation like so `model.embed( channel, LinkRelation.of("channel") )` that does work becasue it does not take in to considuration the `CurieProvider` when you specify the realtion as a paramater this might be by design but IDK. I now just removed the the `LinkRelation` and it now automaticly creates `bp:channelList` which IMO is a good name for the relation. Might need to dig deeper in to this but its out of the scope of this PR so I will leave it like this for now

dusan-turajlic committed:
(baby-project-server) Updates expected result in the test
